[system-probe] Automatically download kernel headers (2)

ISauve opened the pull request:
### What does this PR do?
Follow-up to #7871 containing the changes necessary for building nikos with support for RPM distributions (using `omnibus-nikos`).

~Blocked on https://github.com/DataDog/nikos/pull/7~

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

1. Use vagrant to start & ssh into an RPM-based linux VM (for example https://app.vagrantup.com/generic/boxes/opensuse42)
2. In your `system-probe.yaml` file, enable runtime compilation & set `kernel_header_dirs` to `/dev/null` (this tells the system-probe to check for kernel headers in an incorrect location & forces the download of kernel headers):

```
system_probe_config:
  enabled: true
  log_level: debug
  enable_runtime_compiler: true
  kernel_header_dirs: /dev/null
```

2.  Start the system-probe & verify in the logs that kernel headers get downloaded/validated, and that runtime compilation of the network tracer succeeds:

```
...
| Downloading kernel headers for target distribution [ ... ]
...
| successfully downloaded kernel headers to [...]
| found valid kernel headers at [...]
...
| module: network_tracer started
...
```

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

ISauve committed:
Build the system-probe with support for kernel header downloading on RPM

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -97,3 +97,23 @@ build_clang_arm64:
+build_omnibus-nikos_x64:
+  extends: .build_omnibus-nikos_common
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/omnibus-nikos_x64:$DATADOG_AGENT_NIKOS_BUILDIMAGES
+  tags: ["runner:main", "size:2xlarge"]
```
I believe these `size` tags are outdated and not needed. See #8806

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -97,3 +97,23 @@ build_clang_arm64:
+    - git clone https://github.com/DataDog/omnibus-nikos.git /tmp/omnibus-nikos
+    - cd /tmp/omnibus-nikos
+    - /bin/bash -l -c "bundle install"
+    - /bin/bash -l -c "bundle exec omnibus build nikos"
```
How long does the nikos build take? If it takes awhile, we should make this a manual job, make the artifacts permanent, and remove the job dependency on `build_system-probe-x64`.

ISauve replied:
If I remember correctly, it takes 20-30 minutes - so it's not super long but honestly there isn't a good reason to be running that job for every build anyways, so I'll make it manual.

brycekahle commented on tasks/utils.py:
```diff
@@ -133,6 +173,11 @@ def get_build_flags(
+    # adding nikos libs to the env
+    if nikos_embedded_path is not None:
+        env['PKG_CONFIG_PATH'] = env.get('PKG_CONFIG_PATH', '') + ':' + nikos_embedded_path + '/lib/pkgconfig'
+        env["CGO_LDFLAGS"] = env.get('CGO_LDFLAGS', '') + get_nikos_linker_flags(nikos_embedded_path + '/lib')
```
Is there a way we could include these flags via `#cgo` comments like this? https://github.com/DataDog/datadog-agent/blob/a6dd21d80946c5f8bc4aa8ef25c5c4671ecbbc77/pkg/ebpf/compiler/compiler.go#L6

ISauve replied:
I tried this just to make sure it would work before answering, and for some reason setting the identical linker flags in a #cgo comment instead of here lead to some linker errors (some undefined reference errors & one `error adding symbols: DSO missing from command line` error). These are fixable, so I do still think it would be possible - however, I don't really think it's worth doing.
To make sure that we're building against a statically compiled, specific version of each dependent library, all of the linker flags are in the form `/opt/nikos/embedded/lib/ + library name + .a`. As a result, the #cgo comment which would replace these flags is super verbose & it's hard to see what's actually being included. Also, it has to be placed in a file within the datadog-agent, since adding it to the source files in nikos would make it very difficult to anyone other than us to build/use nikos. This means we'd be throwing a #cgo comment and including "C" into a file which contains no other cgo & isn't associated with any C source files, which works but it rather obfuscates the build process IMO. Finally, it means that the path to the libraries will always be hardcoded, whereas now the linker flags are generated at runtime based on whatever value is passed to the task with `nikos_embedded_path`.

brycekahle replied:
Have you tried specifying the libraries in the form of `-ldnf` (or even `-ldnf.a` if the `.a` extension is required) and then having a single `-L/opt/nikos/embedded/lib`? That should be sufficient. 

> since adding it to the source files in nikos would make it very difficult for anyone other than us to build/use nikos.

Having these in nikos itself, sounds like the correct place for them since they are required to build nikos. If other folks are trying to build from a different directory, they can always use the environment variables to specify alternative directories.

> path to the libraries will always be hardcoded
We hardcode the path to other libraries, so I don't see this as a big deal. As mentioned above, they can use env vars to override.

brycekahle replied:
We chatted offline and are going to tackle this in a future PR.

ISauve committed:
Remove size tag from build_omnibus-nikos_x64

ISauve committed:
Make build_omnibus-nikos a manually triggered job

ISauve committed:
Fix python linting

mx-psi commented on tasks/system_probe.py:
```diff
@@ -88,14 +90,20 @@ def build(
         build_tags.append(BUNDLE_TAG)
     if with_bcc:
         build_tags.append(BCC_TAG)
+    if nikos_embedded_path is not None:
```
```suggestion
    if nikos_embedded_path:
```

mx-psi commented on tasks/utils.py:
```diff
@@ -133,6 +175,11 @@ def get_build_flags(
     )
 
+    # adding nikos libs to the env
+    if nikos_embedded_path is not None:
```
```suggestion
    if nikos_embedded_path:
```

mx-psi reviewed (commented):
Just a couple nits, looks good otherwise

ISauve committed:
Update tasks/system_probe.py

Co-authored-by: Pablo Baeyens <pablo.baeyens@datadoghq.com>

ISauve committed:
Update tasks/utils.py

Co-authored-by: Pablo Baeyens <pablo.baeyens@datadoghq.com>

sgnn7 commented on tasks/utils.py:
```diff
@@ -68,6 +68,47 @@ def get_multi_python_location(embedded_path=None, rtloader_root=None):
+        'lzma',
+        'xml2',
+        'popt',
+        'zstd',
```
nit: Would be nice if these were sorted in some fashion

ISauve replied:
The ordering of these flags is deliberate. If library A uses symbols which are defined in library B, then the flag for library B must appear after library A in order for the symbols to be resolved.

sgnn7 reviewed (approved):
LGTM for the agent-owned files

ISauve committed:
Remove verbose tags from tag commands

ISauve committed:
Remove static libresolv from build flags

ISauve committed:
Include nikos files in agent release packages

ISauve committed:
Update nikos dependency & fix python linting

ISauve commented:
force-pushed to rebase on master in order to solve conflicts

ISauve committed:
Fix system-probe.rb
- add missing mkdir statement
- include all nikos files in order to pass omnibus safety check

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -55,3 +55,25 @@ build_clang_arm64:
+  script:
+    - git clone https://github.com/DataDog/omnibus-nikos.git /tmp/omnibus-nikos
+    - cd /tmp/omnibus-nikos
+    - /bin/bash -l -c "bundle install"
```
Can you explain why this needs a subshell (or why it isn't simply `bundle install`)?

ISauve replied:
It's because of the way that rvm/bundler had to be installed on the `omnibus-nikos_x64` build image.`bundle` is not recognized as a command because it was originally installed this same way (see [here](https://github.com/DataDog/datadog-agent-buildimages/blob/cb0ed517f6a42d7c581ac0c5b584d803969244dc/omnibus-nikos_x64/Dockerfile#L49-L52)). I can't remember the specifics of why it has to be done like that, but FWIW that's how it's done on the other build images as well.

brycekahle commented on omnibus/config/software/system-probe.rb:
```diff
@@ -32,4 +32,11 @@
   copy 'pkg/ebpf/c/COPYING', "#{install_dir}/embedded/share/system-probe/ebpf/"
+
+  if not ENV['NIKOS_PATH'].empty?
+    copy "#{ENV['NIKOS_PATH']}/bin", "#{install_dir}/embedded/nikos/embedded/"
```
I think you need some `mkdir` statements before this. I believe this is causing the CI to fail right now.
